Powerspectrum estimator add threshold

elavallee opened the pull request:
Please review my change to add a threshold to the PowerSpectrum estimator. The optional threshold is named: $\delta$. I updated the docstring, and I was able to pass all the tests locally. However, I did not add any new tests, which is more challenging. For an example of how to use the new feature, please see the code below.
`
```julia
using DynamicalSystems, ComplexityMeasures

N1, N2, a = 101, 100001, 10

for N in (N1, N2)
    println("N: $(N)")
    t = LinRange(0, 2*a*π, N)
    x = sin.(t) # periodic
    y = sin.(t .+ cos.(t/0.5)) # periodic, complex spectrum
    z = sin.(rand(1:15, N) ./ rand(1:10, N)) # random
    w, t = trajectory(Systems.lorenz(), N÷10; Δt = 0.1, Ttr = 100) # chaotic
    w    = w[:, 1]

    for q in (x, y, z, w)
        p, ~ = probabilities_and_outcomes(PowerSpectrum(100.0), q)
        h = entropy(p)
        println("entropy: $(h).")
    end
end
````

elavallee committed:
Adding a new feature to PowerSpectrum estimator
that adds a threshold for "small" amplitudes.

See Issue #104.

elavallee committed:
Functional fixes to the PowerSpectrum threshold
feature.

* Changed the default threshold from 5.0 to 0.0

* Renamed y to amp_squared to make it more clear

* Vectorized the setting of amplitudes to zero

Datseris commented on src/outcome_spaces/power_spectrum.jl:
```diff
@@ -22,14 +23,23 @@ The outcome space `Ω` for `PowerSpectrum` is the set of frequencies in Fourier
     if !(x isa AbstractVector{<:Real})
         throw(ArgumentError("`PowerSpectrum` only works for timeseries input!"))
     end
+    δ = getfield.(Ref(P), (:δ))
```
```suggestion
```

Datseris commented on src/outcome_spaces/power_spectrum.jl:
```diff
@@ -22,14 +23,23 @@ The outcome space `Ω` for `PowerSpectrum` is the set of frequencies in Fourier
     f = FFTW.rfft(x)
-    probs = Probabilities(abs2.(f))
+    amp_squared = abs2.(f)
+    amp_squared[0.0 .< amp_squared  .< δ] .= 0.0
```
```suggestion
    if P.δ > 0
        amp_squared[amp_squared  .< P.δ] .= 0.0
    end
```

Datseris reviewed (commented):
some simplifying and performance comments

elavallee committed:
Merge remote-tracking branch 'origin/main' into powerspectrum_estimator_add_threshold

elavallee committed:
Update src/outcome_spaces/power_spectrum.jl

Co-authored-by: George Datseris <datseris.george@gmail.com>

elavallee committed:
Update src/outcome_spaces/power_spectrum.jl

Co-authored-by: George Datseris <datseris.george@gmail.com>

elavallee committed:
Merge branch 'main' into powerspectrum_estimator_add_threshold

elavallee committed:
Merge branch 'powerspectrum_estimator_add_threshold' of https://github.com/elavallee/ComplexityMeasures.jl into powerspectrum_estimator_add_threshold

elavallee committed:
Updates to logs.

*Updated the minor version number.

*Updated the changelog.

Datseris committed:
Update project version from 3.9.5 to 3.9.0

Datseris committed:
Update CHANGELOG for version 3.9

Updated version number and added a new feature description for the PowerSpectrum estimator.

Datseris committed:
simplify docstring

Datseris committed:
simplify source, remove `new`

Datseris reviewed (changes requested):
I'm sorry @elavallee , I forgot something in my review. There are no tests for this new feature... Can we please add a test? For example testing that the probabilities for the signal `x = cos.(range(0, 2π; length = 10000) .+ 1e-2 .* randn(10000)` with some `δ = 0.1` are all 0 except from a single probability? 

This functionality should be tested. Even though I have no doubt the code is correct as it is, it helps for forwards maintainance if all parts of the code are test covered.

elavallee committed:
Adding a new field for PowerSpectrum.

Added two tests for the new fields of
PowerSpectrum.

Updated CHANGELOG.md and examples.md.

elavallee commented:
Alright, I took a stab at it. I think you can see it now as I pushed it. Let me know what you think. Thanks.

elavallee committed:
Removed the type parameter for the new Bool option

Added an assert checking for a threshold that is
too high

elavallee commented on src/outcome_spaces/power_spectrum.jl:
```diff
@@ -22,8 +23,9 @@ The outcome space `Ω` for `PowerSpectrum` is the set of frequencies in Fourier
-@kwdef struct PowerSpectrum{T<:Real} <: OutcomeSpace
+@kwdef struct PowerSpectrum{T<:Real, X<:Bool} <: OutcomeSpace
     δ::T = 0.0
+    apply_threshold_to_spectrum::X = false
```
I removed it. Also, I found a case where a user could set all probabilities to 0 if the threshold is too high. I added an assert for that. I am not sure what your philosophy is on that.

One other thing, I ran make.jl locally on the main branch. It failed for me. I am not sure that my change to example.md broke it, but let me know if I can help out.

Datseris committed:
Apply suggestions from code review

Datseris commented:
Oh, documentation doesn't build...? Is this related to the PR or other problem?

elavallee committed:
Fixed a misspelling of a function name.

elavallee committed:
Merge branch 'powerspectrum_estimator_add_threshold' of https://github.com/elavallee/ComplexityMeasures.jl into powerspectrum_estimator_add_threshold
